<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Marker Identification &amp; Cluster Annotation Helpers • scCustomize</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Marker Identification &amp; Cluster Annotation Helpers">
<meta property="og:description" content="scCustomize">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">scCustomize</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.6.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">Getting Started</a>
</li>
<li>
  <a href="../articles/Installation.html">Install</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="dropdown-header">Plotting Vignettes</li>
    <li>
      <a href="../articles/Gene_Expression_Plotting.html">Plotting #1: Analysis Plots</a>
    </li>
    <li>
      <a href="../articles/QC_Plots.html">Plotting #2: QC Plots</a>
    </li>
    <li>
      <a href="../articles/Sequencing_QC_Plots.html">Plotting #3: Sequencing QC Plots/Analysis</a>
    </li>
    <li>
      <a href="../articles/Iterative_Plotting.html">Plotting #4: Iterative Plotting Functions</a>
    </li>
    <li>
      <a href="../articles/Color_Palettes.html">Customized Color Palettes &amp; Themes</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Helpers &amp; Utilities</li>
    <li>
      <a href="../articles/Read_and_Write_Functions.html">Read &amp; Write Data Functions</a>
    </li>
    <li>
      <a href="../articles/Markers_and_Cluster_Annotation.html">Marker Identification &amp; Cluster Annotation Helpers</a>
    </li>
    <li>
      <a href="../articles/Statistics.html">Statistics Functions</a>
    </li>
    <li>
      <a href="../articles/Helpers_and_Utilities.html">General Helpers &amp; Utilities</a>
    </li>
    <li>
      <a href="../articles/Misc_Functions.html">Misc Functions</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Other Package Interactivity/Functions</li>
    <li>
      <a href="../articles/LIGER_Functions.html">LIGER Plotting &amp; Functionality</a>
    </li>
    <li>
      <a href="../articles/Cell_Bender_Functions.html">Cell Bender Functionality &amp; Plotting</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">News/Changelog</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../articles/FAQ.html">FAQ</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/samuel-marsh/scCustomize/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="Markers_and_Cluster_Annotation_files/header-attrs-2.11/header-attrs.js"></script><script src="Markers_and_Cluster_Annotation_files/kePrint-0.0.1/kePrint.js"></script><link href="Markers_and_Cluster_Annotation_files/lightable-0.0.1/lightable.css" rel="stylesheet">
<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Marker Identification &amp; Cluster Annotation Helpers</h1>
            
            <h4 class="date">Compiled: November 22, 2021</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/samuel-marsh/scCustomize/blob/master/vignettes/Markers_and_Cluster_Annotation.Rmd"><code>vignettes/Markers_and_Cluster_Annotation.Rmd</code></a></small>
      <div class="hidden name"><code>Markers_and_Cluster_Annotation.Rmd</code></div>

    </div>

    
    
<hr>
<style>
p.caption {
  font-size: 0.9em;
}
</style>
<div id="marker-identification-cluster-annotation" class="section level1">
<h1 class="hasAnchor">
<a href="#marker-identification-cluster-annotation" class="anchor"></a>Marker Identification &amp; Cluster Annotation</h1>
<p>scCustomize has several helper functions to assist with identification of marker genes and annotation of clusters. Load packages &amp; Data</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Load Packages</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org">tidyverse</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat">Seurat</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/samuel-marsh/scCustomize">scCustomize</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/traversc/qs">qs</a></span><span class="op">)</span>

<span class="co"># Load example dataset for tutorial</span>
<span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu">pbmc3k.SeuratData</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/pbmc3k.SeuratData/man/pbmc3k.final.html">pbmc3k.final</a></span></code></pre></div>
<div id="helper-functions-available-are" class="section level2">
<h2 class="hasAnchor">
<a href="#helper-functions-available-are" class="anchor"></a>Helper functions available are:</h2>
<ul>
<li>
<code>Extract_Top_Markers</code>: Extract either a filtered data.frame or vector of gene IDs (with option to export named vector) containing top N markers from data.frame.</li>
<li>
<code>Create_Cluster_Annotation_File</code>: Create simple csv file for use in annotating clusters.</li>
<li>
<code>Pull_Cluster_Annotation</code>: Pull cluster annotation information quickly from data.frame/tibble or annotation csv file.</li>
<li>
<code>Rename_Clusters</code>: Wrapper function to easily rename clusters.</li>
</ul>
<p>We will use pbmc3K dataset from SeuratData package as example for this vignette.</p>
</div>
<div id="percent-difference-in-expression" class="section level2">
<h2 class="hasAnchor">
<a href="#percent-difference-in-expression" class="anchor"></a>Percent Difference in Expression</h2>
<p>If we run <code>FindAllMarkers</code> like normal, this is the output.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Run basic FindAllMarkers DE test</span>
<span class="va">all_markers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindAllMarkers.html">FindAllMarkers</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pbmc</span>, verbose <span class="op">=</span> <span class="cn">F</span><span class="op">)</span>

<span class="va">all_markers</span> <span class="op">&lt;-</span> <span class="va">all_markers</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span>
    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span><span class="op">(</span><span class="va">cluster</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span>
    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html">arrange</a></span><span class="op">(</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/desc.html">desc</a></span><span class="op">(</span><span class="va">avg_log2FC</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span>
    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html">arrange</a></span><span class="op">(</span><span class="va">cluster</span><span class="op">)</span></code></pre></div>
<table class="table table table-bordered table-condensed table-responsive table-striped">
<thead><tr>
<th style="text-align:left;">
</th>
<th style="text-align:right;">
p_val
</th>
<th style="text-align:right;">
avg_log2FC
</th>
<th style="text-align:right;">
pct.1
</th>
<th style="text-align:right;">
pct.2
</th>
<th style="text-align:right;">
p_val_adj
</th>
<th style="text-align:left;">
cluster
</th>
<th style="text-align:left;">
gene
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;">
1
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
1.330040
</td>
<td style="text-align:right;">
0.436
</td>
<td style="text-align:right;">
0.110
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:left;">
Naive CD4 T
</td>
<td style="text-align:left;">
CCR7
</td>
</tr>
<tr>
<td style="text-align:left;">
2
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
1.101679
</td>
<td style="text-align:right;">
0.245
</td>
<td style="text-align:right;">
0.084
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:left;">
Naive CD4 T
</td>
<td style="text-align:left;">
LDLRAP1
</td>
</tr>
<tr>
<td style="text-align:left;">
3
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
1.053259
</td>
<td style="text-align:right;">
0.901
</td>
<td style="text-align:right;">
0.594
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:left;">
Naive CD4 T
</td>
<td style="text-align:left;">
LDHB
</td>
</tr>
<tr>
<td style="text-align:left;">
4
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
1.052644
</td>
<td style="text-align:right;">
0.336
</td>
<td style="text-align:right;">
0.104
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:left;">
Naive CD4 T
</td>
<td style="text-align:left;">
LEF1
</td>
</tr>
<tr>
<td style="text-align:left;">
5
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
1.027161
</td>
<td style="text-align:right;">
0.331
</td>
<td style="text-align:right;">
0.110
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:left;">
Naive CD4 T
</td>
<td style="text-align:left;">
PRKCQ-AS1
</td>
</tr>
</tbody>
</table>
<p>While <code><a href="https://satijalab.org/seurat/reference/FindAllMarkers.html">Seurat::FindAllMarkers</a></code>returns the percent of cells in identity 1 (pct.1) and identity 2 (pct.2) that express a marker it can be helpful to view the difference in these two measures in addition to the values alone.</p>
<p>scCustomize contains helper function: <code>Add_Pct_Diff</code> to add the percent difference between two clusters. <code>Add_Pct_Diff</code> can be used with any output from Seurat, SeuratWrappers, or other DE tests and can be piped with the function itself.</p>
<div id="use-add_pct_diff-piped-with-the-marker-function-" class="section level3">
<h3 class="hasAnchor">
<a href="#use-add_pct_diff-piped-with-the-marker-function-" class="anchor"></a>Use <code>Add_Pct_Diff</code> piped with the marker function.</h3>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># run piped with Seurat command</span>
<span class="va">all_markers_pct</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindAllMarkers.html">FindAllMarkers</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pbmc</span>, verbose <span class="op">=</span> <span class="cn">F</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span>
    <span class="fu"><a href="../reference/Add_Pct_Diff.html">Add_Pct_Diff</a></span><span class="op">(</span><span class="op">)</span>

<span class="va">all_markers_pct</span> <span class="op">&lt;-</span> <span class="va">all_markers_pct</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span>
    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span><span class="op">(</span><span class="va">cluster</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span>
    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html">arrange</a></span><span class="op">(</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/desc.html">desc</a></span><span class="op">(</span><span class="va">avg_log2FC</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span>
    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html">arrange</a></span><span class="op">(</span><span class="va">cluster</span><span class="op">)</span></code></pre></div>
<table class="table table table-bordered table-condensed table-responsive table-striped">
<thead><tr>
<th style="text-align:left;">
</th>
<th style="text-align:right;">
p_val
</th>
<th style="text-align:right;">
avg_log2FC
</th>
<th style="text-align:right;">
pct.1
</th>
<th style="text-align:right;">
pct.2
</th>
<th style="text-align:right;">
p_val_adj
</th>
<th style="text-align:left;">
cluster
</th>
<th style="text-align:left;">
gene
</th>
<th style="text-align:right;">
pct_diff
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;">
1
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
1.330040
</td>
<td style="text-align:right;">
0.436
</td>
<td style="text-align:right;">
0.110
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:left;">
Naive CD4 T
</td>
<td style="text-align:left;">
CCR7
</td>
<td style="text-align:right;">
0.326
</td>
</tr>
<tr>
<td style="text-align:left;">
2
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
1.101679
</td>
<td style="text-align:right;">
0.245
</td>
<td style="text-align:right;">
0.084
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:left;">
Naive CD4 T
</td>
<td style="text-align:left;">
LDLRAP1
</td>
<td style="text-align:right;">
0.161
</td>
</tr>
<tr>
<td style="text-align:left;">
3
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
1.053259
</td>
<td style="text-align:right;">
0.901
</td>
<td style="text-align:right;">
0.594
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:left;">
Naive CD4 T
</td>
<td style="text-align:left;">
LDHB
</td>
<td style="text-align:right;">
0.307
</td>
</tr>
<tr>
<td style="text-align:left;">
4
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
1.052644
</td>
<td style="text-align:right;">
0.336
</td>
<td style="text-align:right;">
0.104
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:left;">
Naive CD4 T
</td>
<td style="text-align:left;">
LEF1
</td>
<td style="text-align:right;">
0.232
</td>
</tr>
<tr>
<td style="text-align:left;">
5
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
1.027161
</td>
<td style="text-align:right;">
0.331
</td>
<td style="text-align:right;">
0.110
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:left;">
Naive CD4 T
</td>
<td style="text-align:left;">
PRKCQ-AS1
</td>
<td style="text-align:right;">
0.221
</td>
</tr>
</tbody>
</table>
</div>
<div id="use-add_pct_diff-post-hoc" class="section level3">
<h3 class="hasAnchor">
<a href="#use-add_pct_diff-post-hoc" class="anchor"></a>Use <code>Add_Pct_Diff</code> post-hoc</h3>
<p><code>Add_Pct_Diff</code> can also simply be run after generating marker data.frame:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">all_markers_pct</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Add_Pct_Diff.html">Add_Pct_Diff</a></span><span class="op">(</span>marker_dataframe <span class="op">=</span> <span class="va">all_markers</span><span class="op">)</span></code></pre></div>
</div>
<div id="use-with-non-seurat-data-frames" class="section level3">
<h3 class="hasAnchor">
<a href="#use-with-non-seurat-data-frames" class="anchor"></a>Use with non-Seurat data.frames</h3>
<p><code>Add_Pct_Diff</code> can be used with any data.frame as long as the column names that specify percent expressed in the 2 groups are provided</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">all_markers_pct</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Add_Pct_Diff.html">Add_Pct_Diff</a></span><span class="op">(</span>marker_dataframe <span class="op">=</span> <span class="va">all_markers</span>, pct.1_name <span class="op">=</span> <span class="st">"percent1"</span>, pct.2_name <span class="op">=</span> <span class="st">"percent2"</span><span class="op">)</span></code></pre></div>
</div>
</div>
<div id="extract-the-top-n-marker-genes-per-cluster-identity-treatment-etc" class="section level2">
<h2 class="hasAnchor">
<a href="#extract-the-top-n-marker-genes-per-cluster-identity-treatment-etc" class="anchor"></a>Extract the top N marker genes per cluster, identity, treatment etc</h2>
<p>scCustomize contains function <code>Extract_Top_Markers</code> to get the top N genes from each cluster (or other group of choice). It has the option to return either a vector of genes or a filtered data.frame.</p>
<div id="returns-a-named-vector-default" class="section level3">
<h3 class="hasAnchor">
<a href="#returns-a-named-vector-default" class="anchor"></a>Returns a Named Vector (Default)</h3>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># By default it returns a named vector of genes (each gene named for cluster it was associated</span>
<span class="co"># with)</span>
<span class="va">top_5</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Extract_Top_Markers.html">Extract_Top_Markers</a></span><span class="op">(</span>marker_dataframe <span class="op">=</span> <span class="va">all_markers</span>, num_genes <span class="op">=</span> <span class="fl">5</span>, rank_by <span class="op">=</span> <span class="st">"avg_log2FC"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">top_5</span>, <span class="fl">10</span><span class="op">)</span></code></pre></div>
<pre><code>##  Naive CD4 T  Naive CD4 T  Naive CD4 T  Naive CD4 T  Naive CD4 T Memory CD4 T 
##       "CCR7"    "LDLRAP1"       "LDHB"       "LEF1"  "PRKCQ-AS1"        "LTB" 
## Memory CD4 T Memory CD4 T Memory CD4 T Memory CD4 T 
##       "AQP3"        "CD2"       "IL32"       "IL7R"</code></pre>
</div>
<div id="return-data-frame-optional" class="section level3">
<h3 class="hasAnchor">
<a href="#return-data-frame-optional" class="anchor"></a>Return data.frame (Optional)</h3>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">top_5_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Extract_Top_Markers.html">Extract_Top_Markers</a></span><span class="op">(</span>marker_dataframe <span class="op">=</span> <span class="va">all_markers</span>, num_genes <span class="op">=</span> <span class="fl">5</span>, data_frame <span class="op">=</span> <span class="cn">TRUE</span>,
    rank_by <span class="op">=</span> <span class="st">"avg_log2FC"</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">top_5_df</span>, <span class="fl">10</span><span class="op">)</span></code></pre></div>
<table class="table table table-bordered table-condensed table-responsive table-striped">
<thead><tr>
<th style="text-align:left;">
</th>
<th style="text-align:right;">
p_val
</th>
<th style="text-align:right;">
avg_log2FC
</th>
<th style="text-align:right;">
pct.1
</th>
<th style="text-align:right;">
pct.2
</th>
<th style="text-align:right;">
p_val_adj
</th>
<th style="text-align:left;">
cluster
</th>
<th style="text-align:left;">
gene
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;">
1
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
1.330040
</td>
<td style="text-align:right;">
0.436
</td>
<td style="text-align:right;">
0.110
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:left;">
Naive CD4 T
</td>
<td style="text-align:left;">
CCR7
</td>
</tr>
<tr>
<td style="text-align:left;">
2
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
1.101679
</td>
<td style="text-align:right;">
0.245
</td>
<td style="text-align:right;">
0.084
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:left;">
Naive CD4 T
</td>
<td style="text-align:left;">
LDLRAP1
</td>
</tr>
<tr>
<td style="text-align:left;">
3
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
1.053259
</td>
<td style="text-align:right;">
0.901
</td>
<td style="text-align:right;">
0.594
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:left;">
Naive CD4 T
</td>
<td style="text-align:left;">
LDHB
</td>
</tr>
<tr>
<td style="text-align:left;">
4
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
1.052644
</td>
<td style="text-align:right;">
0.336
</td>
<td style="text-align:right;">
0.104
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:left;">
Naive CD4 T
</td>
<td style="text-align:left;">
LEF1
</td>
</tr>
<tr>
<td style="text-align:left;">
5
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
1.027161
</td>
<td style="text-align:right;">
0.331
</td>
<td style="text-align:right;">
0.110
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:left;">
Naive CD4 T
</td>
<td style="text-align:left;">
PRKCQ-AS1
</td>
</tr>
<tr>
<td style="text-align:left;">
948
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
1.287053
</td>
<td style="text-align:right;">
0.981
</td>
<td style="text-align:right;">
0.642
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:left;">
Memory CD4 T
</td>
<td style="text-align:left;">
LTB
</td>
</tr>
<tr>
<td style="text-align:left;">
949
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
1.238703
</td>
<td style="text-align:right;">
0.422
</td>
<td style="text-align:right;">
0.110
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:left;">
Memory CD4 T
</td>
<td style="text-align:left;">
AQP3
</td>
</tr>
<tr>
<td style="text-align:left;">
950
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
1.234832
</td>
<td style="text-align:right;">
0.652
</td>
<td style="text-align:right;">
0.244
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:left;">
Memory CD4 T
</td>
<td style="text-align:left;">
CD2
</td>
</tr>
<tr>
<td style="text-align:left;">
951
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
1.208094
</td>
<td style="text-align:right;">
0.948
</td>
<td style="text-align:right;">
0.464
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:left;">
Memory CD4 T
</td>
<td style="text-align:left;">
IL32
</td>
</tr>
<tr>
<td style="text-align:left;">
952
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
1.175375
</td>
<td style="text-align:right;">
0.747
</td>
<td style="text-align:right;">
0.325
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:left;">
Memory CD4 T
</td>
<td style="text-align:left;">
IL7R
</td>
</tr>
</tbody>
</table>
</div>
<div id="use-with-non-seurat-data-frames-1" class="section level3">
<h3 class="hasAnchor">
<a href="#use-with-non-seurat-data-frames-1" class="anchor"></a>Use with non-Seurat data.frames</h3>
<p><code>Extract_Top_Markers</code> uses the data.frame output of <code><a href="https://satijalab.org/seurat/reference/FindAllMarkers.html">Seurat::FindAllMarkers</a></code> by default but can be used with any data.frame by supplying few extra parameters. Let’s use the function on the output from <a href="https://github.com/immunogenomics/presto">presto package</a> and order the results by auc instead of log2FC.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Read in pre-computed data.frame results</span>
<span class="va">presto_markers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/qs/man/qread.html">qread</a></span><span class="op">(</span><span class="st">"assets/presto_markers.qs"</span><span class="op">)</span>

<span class="co"># Extract top markers</span>
<span class="va">top_5_presto</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Extract_Top_Markers.html">Extract_Top_Markers</a></span><span class="op">(</span>marker_dataframe <span class="op">=</span> <span class="va">presto_markers</span>, num_genes <span class="op">=</span> <span class="fl">5</span>, group_by <span class="op">=</span> <span class="st">"cluster"</span>,
    rank_by <span class="op">=</span> <span class="st">"auc"</span>, gene_column <span class="op">=</span> <span class="st">"gene"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">top_5_presto</span>, <span class="fl">10</span><span class="op">)</span></code></pre></div>
<pre><code>##  Naive CD4 T  Naive CD4 T  Naive CD4 T  Naive CD4 T  Naive CD4 T Memory CD4 T 
##      "RPS12"      "RPS27"       "RPS6"      "RPL32"      "RPS14"        "LTB" 
## Memory CD4 T Memory CD4 T Memory CD4 T Memory CD4 T 
##       "IL32"       "LDHB"       "CD3D"       "IL7R"</code></pre>
</div>
</div>
<div id="pairing-extract_top_markers-with-iterate_featureplot_sccustom" class="section level2">
<h2 class="hasAnchor">
<a href="#pairing-extract_top_markers-with-iterate_featureplot_sccustom" class="anchor"></a>Pairing <code>Extract_Top_Markers</code> with <code>Iterate_FeaturePlot_scCustom</code>
</h2>
<p>The advantage of exporting a named vector from <code>Extract_Top_Markers</code> is that <code>Iterate_FeaturePlot_scCustom</code> can detect the names and either add those to the title of the plot (if <code>single_pdf = TRUE</code>) or the file names if <code>FALSE</code>.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/Iterate_FeaturePlot_scCustom.html">Iterate_FeaturePlot_scCustom</a></span><span class="op">(</span>seurat_object <span class="op">=</span> <span class="va">pbmc</span>, gene_list <span class="op">=</span> <span class="va">top_5_presto</span>, single_pdf <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="../docs/articles/assets/Iterate_named_plots.png" alt="*Example plot showing the gene name and additional name of the plot.*" width="49%"><p class="caption">
<em>Example plot showing the gene name and additional name of the plot.</em>
</p>
</div>
</div>
<div id="create-simple-annotation-files" class="section level2">
<h2 class="hasAnchor">
<a href="#create-simple-annotation-files" class="anchor"></a>Create simple annotation files</h2>
<p>Now for a simple helper we can use <code>Create_Cluster_Annotation_File</code> to create simple csv file to use for annotation.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/Create_Cluster_Annotation_File.html">Create_Cluster_Annotation_File</a></span><span class="op">(</span>file_path <span class="op">=</span> <span class="st">"data/"</span>, file_name <span class="op">=</span> <span class="st">"ExpXX_Round01_Annotation"</span><span class="op">)</span></code></pre></div>
<p>Here is an example that has already been filled out (using Excel, Numbers, or other editor program).</p>
<table class="table table table-bordered table-condensed table-responsive table-striped">
<thead><tr>
<th style="text-align:left;">
</th>
<th style="text-align:right;">
cluster
</th>
<th style="text-align:left;">
cell_type
</th>
<th style="text-align:left;">
sub_type
</th>
<th style="text-align:left;">
notes
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;">
1
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:left;">
oligo
</td>
<td style="text-align:left;">
</td>
<td style="text-align:left;">
Olig2+
</td>
</tr>
<tr>
<td style="text-align:left;">
2
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:left;">
micro
</td>
<td style="text-align:left;">
</td>
<td style="text-align:left;">
Proliferating
</td>
</tr>
<tr>
<td style="text-align:left;">
3
</td>
<td style="text-align:right;">
2
</td>
<td style="text-align:left;">
micro
</td>
<td style="text-align:left;">
</td>
<td style="text-align:left;">
</td>
</tr>
<tr>
<td style="text-align:left;">
4
</td>
<td style="text-align:right;">
3
</td>
<td style="text-align:left;">
oligo
</td>
<td style="text-align:left;">
</td>
<td style="text-align:left;">
Newly formed
</td>
</tr>
<tr>
<td style="text-align:left;">
5
</td>
<td style="text-align:right;">
4
</td>
<td style="text-align:left;">
oligo
</td>
<td style="text-align:left;">
</td>
<td style="text-align:left;">
</td>
</tr>
</tbody>
</table>
</div>
<div id="extract-annotation-information" class="section level2">
<h2 class="hasAnchor">
<a href="#extract-annotation-information" class="anchor"></a>Extract annotation information</h2>
<p>There is also complementary function to pull information back out of the this (or other) csv files or data.frames regarding clustering annotation.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">annotation_info</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Pull_Cluster_Annotation.html">Pull_Cluster_Annotation</a></span><span class="op">(</span>annotation <span class="op">=</span> <span class="st">"assets/ExpXX_Round01_Annotation_Completed.csv"</span><span class="op">)</span></code></pre></div>
<p>By default the function expects data.frame or file in same format as <code>Create_Cluster_Annotation_File</code> but has optional parameters that can be adjusted to file/data.frame with different column headers (see <code>cluster_name_col</code> and <code>cell_type_col</code>).</p>
<div id="pull_cluster_annotation-results" class="section level3">
<h3 class="hasAnchor">
<a href="#pull_cluster_annotation-results" class="anchor"></a><code>Pull_Cluster_Annotation</code> results</h3>
<p>The result of <code>Pull_Cluster_Annotation</code> is a list that contains:</p>
<ul>
<li>One entry for each unique cell type. The entry is named for the cell type and is a vector of the numbered clustered that correspond to that cell type.</li>
<li>One entry called “new_cluster_idents” containing a vector of idents in order that can be used for renaming clusters either using <code>Rename_Clusters</code> function (see below) or manually.</li>
<li>One entry containing vector of cell sub-types (if present).
<div class="figure" style="text-align: center">
<img src="../docs/articles/assets/annotation_info.png" alt="*Example output of `Pull_Cluster_Annotation*" width="831"><p class="caption">
<em>Example output of `Pull_Cluster_Annotation</em>
</p>
</div>
</li>
</ul>
</div>
</div>
<div id="renaming-clusters" class="section level2">
<h2 class="hasAnchor">
<a href="#renaming-clusters" class="anchor"></a>Renaming Clusters</h2>
<p>Finally can use <code>Rename_Clusters</code> to easily rename and set active.ident in Seurat Object. This is simple wrapper around Seurat commands but adds additional checks/warnings and is simple one-liner.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">obj_renamed</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Rename_Clusters.html">Rename_Clusters</a></span><span class="op">(</span>seurat_object <span class="op">=</span> <span class="va">obj</span>, new_idents <span class="op">=</span> <span class="va">annotation_info</span><span class="op">$</span><span class="va">new_cluster_idents</span><span class="op">)</span></code></pre></div>
<p><code>Rename_Clusters</code> can take input from either <code>Pull_Cluster_Annotation</code> “new_cluster_idents” or any correctly ordered vector of new idents.</p>
<ul>
<li>If vector is not yet named (with the current levels of the Seurat Object) then <code>Rename_Clusters</code> will perform that step. If already named this step will be omitted.</li>
<li>
<code>Rename_Clusters</code> contains optional parameter to create new column in meta.data slot to store the old idents. Simply supply desired meta.data column name to <code>meta_col_name</code> parameter.</li>
</ul>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Samuel Marsh.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
